### DISCUSSION TITLE
File import - Hierarchical Data Format (HDF) Files (.h5)

### DISCUSSION THREAD

**Discussion by alhadjil** [JacobBumgarner/VesselVio#25] [Ideas]
Hi there,

thanks for this interactive tool! Looking forward to explore all of its potential.

I have stacked pics taken with LSFM saved as a hierarchical data format (HDF) files (.h5) and I'm wondering if VesselVio can be tunned to take such files directly as import. If not, how would you recommend to proceed?

Thanks!

**Comment by JacobBumgarner** [comment]
Hi @alhadjil! Thanks for checking out VesselVio. 

Honestly, I think it would be best to implement compatibility with `.h5` files. This seems simple enough to me, as there are numerous packages available that are compatible with HDF files (`h5py` and even `pandas`).

Two quick questions:
1. Do you have a small `.h5` file that you'd be willing to share with me for testing purposes?
2. Are you using VesselVio from the terminal? If so, you will be able to use `.h5` input as soon as I update the `image_processing.py` code. The 1.1.2 release is still a week or two away, and this feature will likely be included in that update.

**Reply by JacobBumgarner** [reply]
It also seems like `.h5` files store header information, and the exact name of the image in the `.h5` file is needed in order to access the file from the hf data frame. [See docs here](https://docs.h5py.org/en/stable/quick.html).

This means that if users are planning on loading `.h5` files, I'll probably have to place a specific call to ask them what the name of the header is that points to the dataset. 

Does this sound accurate to you?

**Reply by alhadjil** [reply]
Hey,

thanks for the prompt reply!

Unfortunately, my files are too big to upload. I'm a newbie in programming, so I was excited to test the interactive version `.exe` of VesselVio. I could add the updated `image_processing.py` in my folder for `VesselVio.exe` use, no?

**Reply by JacobBumgarner** [reply]
If you'd be willing to slice a small volume of your dataset (<25mb) and then share it with me for testing purposes, that would be great. If not, no worries. Ideally it would be nice for me to have a 'real world' dataset for testing rather than something that I just generate locally.

> I could add the updated `image_processing.py` in my folder for VesselVio.exe use, no?

Unfortunately, that's not quite how the bundled application works. The `library` folder is only included because I explicitly add it during bundling, otherwise those files wouldn't be present. 

I will add this to the list of features that I would like to include in the v1.1.2 update :) In the meantime, see below for our temporary solution (`.h5` -> `.nii` conversion).

**Comment by alhadjil** [comment]
What about converting `.h5` in `.nii`?

**Reply by JacobBumgarner** [reply]
As a temporary solution. `.h5` conversion to `.nii` would certainly work!

**Reply by alhadjil** [reply]
Any simple way to do so?

**Reply by JacobBumgarner** [reply]
Please let me know if you'd like me to write a quick snippet of code to convert your `.h5` to `.nii` in the meantime.

**Reply by JacobBumgarner** [reply]
Crazy timing lol.

Do you have a python environment set up on your computer?

If so, I can write a script for you. 

If you're not comfortable with python, you can try FIJI too!
Make sure the [HDF5](https://lmb.informatik.uni-freiburg.de/resources/opensource/imagej_plugins/hdf5.html) plugin is installed, and then save the file using the [NIfTI](https://imagej.nih.gov/ij/plugins/nifti.html) plugin using File>Save As>NIfTI-1

**Reply by alhadjil** [reply]
I could try in python. Thanks for your support!!!

**Reply by alhadjil** [reply]
or fiji ^^

**Reply by JacobBumgarner** [reply]
Cool. As soon as I'm out of this meeting I'll get something written for you ! ~1h

**Reply by alhadjil** [reply]
I'm working it out with Fiji for the moment, let's see how it goes...

**Reply by JacobBumgarner** [reply]
Sounds good. Just shoot me a note if you need help!

**Reply by JacobBumgarner** [reply]
Ok - maybe we can start from here so that I can figure out the shape of your volume.

You'll need to make sure that you have numpy and h5py installed. Since you're using a JupiterLab notebook, the following should work to make sure that they're installed:

```python
import sys
!{sys.executable} -m pip install numpy h5py
```

After they're installed, try running this. Find the keys of your h5 file, then replace `'KEY_HERE'` with the appropriate key that points to your volume.

```python
import h5py
import numpy as np

filename = "/Users/jacobbumgarner/Desktop/NEONDSTowerTemperatureData.hdf5"

hfile = h5py.File(filename, 'r')

file_keys = list(hfile.keys())
print (f"H5 file keys: {file_keys}")

volume = np.asarray(hfile['KEY_HERE'])

print (f"Volume shape: {volume.shape}")
print (f"Volume data type: {volume.dtype}")
```

Could you then share the print output of your file keys, the volume shape, and the volume data type with me?
